### PR TITLE
feat(Auth): Creating a unit test harness for cross platform API tests

### PR DESCRIPTION
**Pull request by harsh62**
*Issue #, if available:*

*Description of changes:*

*Check points: (check or cross out if not relevant)*

- [x] Added new tests to cover change, if needed
- [x] Build succeeds with all target using Swift Package Manager
- [x] All unit tests pass
- [x] All integration tests pass
- [x] Security oriented best practices and standards are followed (e.g. using input sanitization, principle of least privilege, etc)
- [ ] Documentation update for the change if required
- [x] PR title conforms to conventional commit style
- [ ] If breaking change, documentation/changelog update with migration instructions

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: AmplifyPlugins/Auth/Tests/AWSCognitoAuthPluginUnitTests/TestHarness/AmplifyAuthCognitoPluginTests.swift [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+                        }
+                    }
+                    expectation.fulfill()
+                } catch {
+                    XCTFail("Reset password API should throw AuthError")
+                    expectation.fulfill()
+                }
+            }
+            wait(for: [expectation], timeout: apiTimeout)
+        }
```
Comment (royjit) [inline AmplifyPlugins/Auth/Tests/AWSCognitoAuthPluginUnitTests/TestHarness/AmplifyAuthCognitoPluginTests.swift:87]:
```suggestion
          do {
                    let result = try await apiCall()
                    XCTAssertEqual(expectedOutput, result)

                } catch let error as AuthError {
                   XCTAssertEqual(expectedOutput, result)
                } catch {
                    XCTFail("Reset password API should throw AuthError")
                }
```


Does this work?

File: AmplifyPlugins/Auth/Tests/AWSCognitoAuthPluginUnitTests/TestHarness/AmplifyAPIDecoding/AmplifyAPIDecodingHelper.swift [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+//
+
+@testable import Amplify
+@testable import AWSCognitoAuthPlugin
+import Foundation
+
+
+struct AmplifyAPIDecodingHelper {
+
+    static func decodeAmplifyAPI(
```
Comment (royjit) [inline AmplifyPlugins/Auth/Tests/AWSCognitoAuthPluginUnitTests/TestHarness/AmplifyAPIDecoding/AmplifyAPIDecodingHelper.swift:15]:
```suggestion
    static func decode(
```

File: AmplifyTestCommon/TestCommonConstants.swift [hunk 3]
```diff
@@ -6,6 +6,6 @@
 //
 
 import Foundation
-class TestCommonConstants {
-    static let networkTimeout = TimeInterval(10)
+public class TestCommonConstants {
+    public static let networkTimeout = TimeInterval(10)
```
Comment (royjit) [inline AmplifyTestCommon/TestCommonConstants.swift:10]: Let us have a auth plugin specific time interval and reduce the time to maybe 2 or 3. The time requirement for test can vary across plugin and it is better to have plugin specific values.

Comment (harsh62) [reply AmplifyTestCommon/TestCommonConstants.swift:10]: Not using this constants for the time being..

File: AmplifyTestCommon/Helpers/XCTestCase+AsyncTesting.swift [hunk 4]
```diff
@@ -23,7 +23,7 @@ extension XCTestCase {
     ///   
     /// - Returns:The result of successfuly running `action`, or `nil` if it threw an error.
     @discardableResult
-    func wait<T>(with expectation: AsyncExpectation,
+    public func wait<T>(with expectation: AsyncExpectation,
```
[the remark this answers is not in the document]
Comment (harsh62) [reply AmplifyTestCommon/Helpers/XCTestCase+AsyncTesting.swift:26]: Don't need it.. Still keeping it public, we would need it in out integration tests.

**Review by royjit — APPROVED** [review]

**Pull request merged**